libraries/aws_cloudwatch_alarm.rb

```diff
@@ -0,0 +1,104 @@
+require 'aws_conn'
+
+class AwsCloudwatchAlarm < Inspec.resource(1)
+  name 'aws_cloudwatch_alarm'
+  desc <<-EOD
+  # Look for a specific alarm
+  aws_cloudwatch_alarm(
+    metric: 'my-metric-name',
+    metric_namespace: 'my-metric-namespace',
+  ) do
+    it { should exist }
+  end
+  EOD
+
+  attr_reader :exists, :alarm_name, :metric_name, :metric_namespace, :alarm_actions
+  def initialize(opts)
+    # Validates and sets instance variables
+    validate_resource_params(opts)
+    search
+  end
+
+  def exists?
+    @exists
+  end
+
+  private
+
+  def validate_resource_params(raw_params)
+    unless raw_params.is_a? Hash
+      raise ArgumentError, "Resource params should be passed using \"key: 'value'\" format."
+    end
+
+    # Currently you must specify exactly metric_name and metric_namespace
+    [:metric_name, :metric_namespace].each do |param|
+      raise ArgumentError, "Missing resource param #{param}" unless raw_params.key?(param)
+      instance_variable_set(:"@#{param}", raw_params.delete(param))
+    end
+
+    # Any leftovers are unwelcome
+    unless raw_params.empty? # rubocop:disable Style/GuardClause
+      raise ArgumentError, "Unrecognized resource param '#{raw_params.keys.first}'"
+    end
+  end
+
+  def search
+    aws_alarms = Backend.create.describe_alarms_for_metric(
+      metric_name: @metric_name,
+      namespace: @metric_namespace,
+    )
+    if aws_alarms.metric_alarms.empty?
+      @exists = false
+    elsif aws_alarms.metric_alarms.count > 1
+      alarms = aws_alarms.metric_alarms.map(&:alarm_name)
+      raise 'More than one Cloudwatch Alarm was matched. Try using ' \
+        "more specific resource parameters. Alarms matched: #{alarms.join(', ')}"
+    else
+      @alarm_actions = aws_alarms.metric_alarms.first.alarm_actions
+      @alarm_name = aws_alarms.metric_alarms.first.alarm_name
+      @exists = true
+    end
+  end
+
+  class Backend
```

Comment:
I am not 100% I am understanding the reason for that abstraction. Can you share your reasoning?

Reply:
In the unit tests, I need to be able to have something that acts like AWS, but actually isn't.  A typical unit test will subclass SomeResource::Backend, then implement the interface, typically using lookup tables to provide dummy data for the tests.  The unit test then calls SomeResource::Backend.select(MyDummyBackend), and we've effectively swapped the provider.

I use the same pattern in #112 and #116.  Other solutions are possible; if we don't like this one, we should get together and design another.

Some resources were mocking the backend using classes passed in through the constructor (#110) which seemed more problematic, since the constructor is part of the public interface of the resource used by control authors.